aten/src/ATen/native/sparse/cuda/SparseMatMul.cu

```diff
@@ -112,21 +113,15 @@ struct csrMatrixRef {
       int* csr_pointers,
       scalar_t* csr_values,
       int nnz,
-      const std::vector<int>& size)
+      const std::vector<int>& size,
+      const at::ScalarType& scalar_type)
       : csr_indices_{csr_indices},
         csr_pointers_{csr_pointers},
         csr_values_{csr_values},
         nnz_{nnz},
         size_{size} {
     #if IS_CUSPARSE11_AVAILABLE()
-      cudaDataType cuda_data_type;
-      if ( std::is_same<float, scalar_t>::value ) {
-        cuda_data_type = CUDA_R_32F;
-      } else if ( std::is_same<double, scalar_t>::value) {
-        cuda_data_type = CUDA_R_64F;
-      } else {
-        TORCH_CHECK(false, "Tensor types must be either float32 or float64");
-      }
+      cudaDataType cuda_data_type = at::cuda::toCudaDataType(scalar_type);
```

Comment:
previous approach was nice because it didn't require redundant `scalar_type` argument when you are already templated on `scalar_t`. Can you template `toCudaDataType` on `scalar_t` and use an approach that was used before in the body? 
```
template <typename scalar_t>
cudaDataType() {
    if (std::is_same<float, scalar_t>::value) {
       return CUDA_R_32F;
    } else { 
    etc
    }
}
```

Reply:
I changed the function to be templated on `scalar_t` and I used partial specialization instead of runtime if-else. Unfortunately `if constexpr` is available only for C++17 and `at::guts::if_constexpr` is difficult to use for many nested if-else conditions.

Reply:
Even without constexpr if compiler is evaluating `if..else` function at compile time https://godbolt.org/z/q4K4nG1do